### PR TITLE
video_core: Take factors into account with min/max blending functions

### PR DESCRIPTION
**Pull request by GPUCode**
This PR emulates a difference between PICA and OpenGL blending behaviour, namely with min/max blending. OpenGL (and Vulkan) don't take the source and destination factors into account with these functions while PICA does, which was hwtested and confimed by Subv [here](https://github.com/citra-emu/citra/issues/2684#issuecomment-306555172)

The software renderer implementation is trivial while for OpenGL we employ a variety of methods depending on the platform. For NVIDIA/AMD gpus the backend uses GL_NV_blend_minmax_factor and GL_AMD_blend_minmax_factor respectively which provide the features we need essentially for free

In the absense of said extensions blending needs to be manually emulated in the fragment shaders. The host blending is configured to be passthrough so the fragment shader outputs are not modified further. The dest factor is retrieved either using the ouput color (GL_EXT_shader_framebuffer_fetch supported on Intel iGPUs and most adreno chips) or gl_LastFragColorARM (GL_ARM_shader_framebuffer_fetch supported on vast majority of Mali gpus)

I have purposly avoided modifying the vulkan shader generators as the vulkan implementation for this feature will be different than opengl and likely use renderpass feedback loops with VK_ARM_rasterization_order_attachment_access for mobile. Testing is required for the following platforms:

- [x] NVIDIA
- [x] AMD
- [x] Intel
- [x] Qcom
- [x] Mali  

Citra (master) | Citra (PR)
-|-
before | after
shin_before | shin_after

Fixes:
https://github.com/citra-emu/citra/issues/2684
https://github.com/citra-emu/citra/issues/2104
https://github.com/citra-emu/citra/issues/2272
https://github.com/citra-emu/citra/issues/5775
https://github.com/citra-emu/citra/issues/5052

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6925)

### CODE REVIEW

File: src/video_core/renderer_opengl/gl_shader_gen.cpp [hunk 1]
```diff
@@ -1224,6 +1316,11 @@ ShaderDecompiler::ProgramResult GenerateFragmentShader(const PicaFSConfig& confi
     if (separable_shader && !GLES) {
         out += "#extension GL_ARB_separate_shader_objects : enable\n";
     }
+    if (config.state.has_ext_framebuffer_fetch) {
+        out += "#extension GL_EXT_shader_framebuffer_fetch : enable\n";
+    } else if (config.state.has_arm_framebuffer_fetch) {
+        out += "#extension GL_ARM_shader_framebuffer_fetch : enable\n";
+    }
```
Comment (Steveice10) [inline src/video_core/renderer_opengl/gl_shader_gen.cpp:1323]: You should be able to simplify state here by using ifdefs in the shader code for these extensions.

https://stackoverflow.com/a/61919013

File: src/video_core/renderer_opengl/gl_shader_gen.cpp [hunk 2]
```diff
@@ -1216,6 +1246,68 @@ float ProcTexNoiseCoef(vec2 x) {
...
 
+static void WriteBlending(std::string& out, const PicaFSConfig& config) {
+    if (!config.state.rgb_blend.emulate_blending && !config.state.alpha_blend.emulate_blending)
+        [[likely]] {
+        return;
+    }
+
+    using BlendFactor = Pica::FramebufferRegs::BlendFactor;
+    const std::string frag_data =
+        config.state.has_ext_framebuffer_fetch ? "gl_LastFragData" : "gl_LastFragColorARM";
```
Comment (Steveice10) [inline src/video_core/renderer_opengl/gl_shader_gen.cpp:1257]: Ditto here, can use extension ifdef's for this and eliminate the generator state.

File: src/video_core/shader/shader_uniforms.cpp [hunk 3]
```diff
@@ -69,6 +69,9 @@ layout ({}std140) uniform shader_data {{
     vec3 tex_lod_bias;
     vec4 tex_border_color[3];
     vec4 clip_coef;
+#ifndef VULKAN
+    vec4 blend_color;
+#endif
```
Comment (Steveice10) [inline src/video_core/shader/shader_uniforms.cpp:74]: No reason to ifdef this out, it's part of the data regardless.

File: src/video_core/renderer_opengl/gl_shader_gen.cpp [hunk 4]
```diff
@@ -60,7 +61,8 @@ out gl_PerVertex {
     return out;
 }
 
-PicaFSConfig PicaFSConfig::BuildFromRegs(const Pica::Regs& regs, bool use_normal) {
+PicaFSConfig PicaFSConfig::BuildFromRegs(const Pica::Regs& regs, const Driver& driver,
```
Comment (Steveice10) [inline src/video_core/renderer_opengl/gl_shader_gen.cpp:64]: With state simplified as in other comments, would rather this just be a bool flag than having to pass the driver in, given I'm working on unifying the generators and this won't work then.

**Review by Steveice10 — COMMENTED** [review]
General comment I have is that doing a lot of the state management and support checking in GL-specific code is going to make this more difficult for my work unifying the shader generators, would prefer if those parts were made more generalized even if we're probably going to turn it off for shaders Vulkan asks for.

File: src/video_core/renderer_opengl/gl_shader_gen.cpp (hunk 4 above)
Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_shader_gen.cpp]: I think when the generators are refactored we should use a separate structure for extensions like yuzu does: https://github.com/yuzu-emu/yuzu/blob/master/src/shader_recompiler/profile.h so the config won't need to deal with them

**Comment by Dragios** [comment]
Should close #5775 as well since it is basically a duplicate issue of #2272

**Comment by LOuroboros** [comment]
I don't think it's right to say that #5775 is a duplicate of #2272 since it was a dedicated ticket for SMT IV and not SMT IVA, but regardless, this also closes #5052, the actual ticket that #5775 is a duplicate of 🥳 !

https://github.com/citra-emu/citra/assets/4485172/e62cbbcb-603c-401e-a589-c67e731f4b51

**Comment by GPUCode** [comment]
I've confirmed the PR is working now on both qcom and mali devices, now only intel needs testing

**Comment by sukerokushin** [comment]
Intel enters battle and renders the effect correctly. [Here's the log just in case](https://github.com/citra-emu/citra/files/12448575/citra_log.txt)

image

**Comment by Steveice10** [comment]
Conflicts with https://github.com/citra-emu/citra/pull/6512 in canary (I think over something trivial like a header include being removed there?)

**Comment by GPUCode** [comment]
Removed conflicting code in the vulkan PR

**Review by Steveice10 — APPROVED** [review]

**Comment by Steveice10** [comment]
Out of curiosity, what particular advantages do the extensions supported here have over the new fallback path implemented?

**Comment by GPUCode** [comment]
The fallback is only guaranteed to work on desktop under the presence of [ARB_texture_barrier](https://registry.khronos.org/OpenGL/extensions/ARB/ARB_texture_barrier.txt) which relaxes the restrictions on feedback loops. The 4.6 [specification](https://registry.khronos.org/OpenGL/specs/gl/glspec46.core.pdf#section.9.3) mentions this use case specifically:

```
Specifically, the values of rendered fragments are undefined ... unless any of the following
exceptions apply:

• There is only a single read and write of each texel, and the read is in
the fragment shader invocation that writes the same texel (e.g. using
texelFetch2D(sampler, ivec2(gl_FragCoord.xy), 0);)
```

The OpenGL ES [specification](https://registry.khronos.org/OpenGL/specs/es/3.2/es_spec_3.2.pdf) on the other hand mentions that doing this produces undefined values, so we need to extension to guarantee correctness

**Comment by Steveice10** [comment]
Thanks. From what I can tell though the fallback shader code does not seem to be gated by anything, will this interfere with using the NV/AMD extensions by blending twice?

**Comment by GPUCode** [comment]
Yeah you're right, corrected that.

**Comment by megamitensei2** [comment]
Greetings! not a big tech guy so I was wondering if you could explain how to get this working in layman's terms

**Comment by rlpowell** [comment]
I just tested this on Nightly and it works! Sooo happy, thank you!

I might try to backport it to the last nightly that can be compiled for Mac M1, but ... there's a *lot* of changes between then and now so I might wait for the MacOS problems to get fixed first.

**Comment by GPUCode** [comment]
> I just tested this on Nightly and it works! Sooo happy, thank you!
> 
> I might try to backport it to the last nightly that can be compiled for Mac M1, but ... there's a _lot_ of changes between then and now so I might wait for the MacOS problems to get fixed first.

Canary works pretty well on MacOS with @Steveice10's fixes. This fix is also available on Canary

**Comment by rlpowell** [comment]
OK, so I grabbed the unified source ball from https://github.com/citra-emu/citra-canary/releases/tag/canary-2607 ; it does in fact compile on M1 hurray.  If I try to pick OpenGL from the graphics I get this:

no_opengl

If I use Vulkan like so:

vulkan

, your fix does not appear.  I thought your fix  was OpenGL specific, so, not super surprised.

**Comment by rlpowell** [comment]
Sorry, when I say "your fix does not appear", I mean I get the black boxes, not that it's not in the code (which I didn't actually check).

**Comment by GPUCode** [comment]
Oh yeah this isn't implemented in vulkan yet, I forgot that for a sec, sorry for wasting time :p
This will be added to vulkan as well eventually but I'm not sure if it will be in a separate PR or in the current one

**Comment by rlpowell** [comment]
Still awesome and I appreciate you.  :)  And the whole team.

**Comment by FonsyCat** [comment]
Just wanting to thank you so much for this! After waiting so long, a lot of people will be happy to play their games without the bugs all the time ♥ I already tested it and it works perfectly!
